aws/ecs_cluster.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+const firstSeenTagKey = "cloud-nuke-first-seen"
+
+// Tag an ECS cluster identified by the given cluster ARN with a tag that has the given key and value
+func tagEcsCluster(awsSession *session.Session, clusterArn *string, tagKey string, tagValue string) error {
+	svc := ecs.New(awsSession)
+	input := &ecs.TagResourceInput{
+		ResourceArn: clusterArn,
+		Tags: []*ecs.Tag{
+			{
+				Key:   aws.String(tagKey),
+				Value: aws.String(tagValue),
+			},
+		},
+	}
+
+	_, err := svc.TagResource(input)
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+	return nil
+}
+
+func getAllEcsClustersOlderThan(awsSession *session.Session, region string, excludeAfter time.Time) ([]*string, error) {
+	awsSession, err := session.NewSession(&awsgo.Config{
+		Region: awsgo.String(region),
+	})
+
+	clusterArns, err := getAllEcsClusters(awsSession)
+	if err != nil {
+		logging.Logger.Errorf("Error getting all ECS clusters")
+		return nil, err
+	}
+
+	var filteredEcsClusters []*string
+	for _, clusterArn := range clusterArns {
+
+		firstSeenTime, err := getClusterTag(awsSession, clusterArn, firstSeenTagKey)
+		if err != nil {
+			logging.Logger.Errorf("Error getting the `cloud-nuke-first-seen` tag for ECS cluster with ARN %s", aws.StringValue(clusterArn))
+			return nil, err
+		}
+
+		if firstSeenTime.IsZero() {
+			err := tagEcsCluster(awsSession, clusterArn, firstSeenTagKey, time.Now().UTC().String())
+			if err != nil {
+				logging.Logger.Errorf("Error tagigng the ECS cluster with ARN %s", aws.StringValue(clusterArn))
+				return nil, err
+			}
+		}
+
+		if excludeAfter.After(firstSeenTime) {
+			filteredEcsClusters = append(filteredEcsClusters, clusterArn)
+		}
+	}
+
+	return filteredEcsClusters, nil
+}
+
+func nukeEcsClusters(awsSession *session.Session, ecsClusterArns []*string) error {
+	svc := ecs.New(awsSession)
+
+	numNuking := len(ecsClusterArns)
+
+	if numNuking == 0 {
+		logging.Logger.Infof("No ECS clusters to nuke in region %s", aws.StringValue(awsSession.Config.Region))
+		return nil
+	}
+
+	logging.Logger.Infof("Deleting %d ECS clusters in region %s", numNuking, aws.StringValue(awsSession.Config.Region))
+
+	var nukedEcsClusters []*string
+	for _, clusterArn := range ecsClusterArns {
+		params := &ecs.DeleteClusterInput{
+			Cluster: clusterArn,
+		}
+		_, err := svc.DeleteCluster(params)
+
+		if err != nil {
+			logging.Logger.Errorf("Error, failed to delete cluster with ARN %s", aws.StringValue(clusterArn))
+		} else {
+			logging.Logger.Infof("Success, deleted cluster: %s", aws.StringValue(clusterArn))
+			nukedEcsClusters = append(nukedEcsClusters, clusterArn)
+		}
+	}
+
+	numNuked := len(nukedEcsClusters)
+	logging.Logger.Infof("[OK] %d of %d ECS cluster(s) deleted in %s", numNuked, numNuking, aws.StringValue(awsSession.Config.Region))
+
+	return nil
+}
+
+func getClusterTag(awsSession *session.Session, clusterArn *string, tagKey string) (time.Time, error) {
```

Comment:
A question about this function. Based on the function name and signature, it looks like a generic "get tag" function that does not make an assumption about what the tag key/value is. However, the implementation is specific to the "first seen" tag - e.g. in the line 122, we are assuming that the `tag.Value` can be parsed into a `time` structure. Is this a concern? Should we refactor, so that there is a generic `getClusterTag` function, and there a more specific one that parses the tag value? (Or at least rename this existing function to suggest that it is opinionated about what kind of tag it is getting?)

Reply:
Ah, big +1 on that. Renaming the function prob makes the most sense for now!

Reply:
Absolutely - I can see it going wrong in future, so will rename it 👍 Thanks @infraredgirl :)